crates/optimism/flashblocks/src/service.rs

```diff
@@ -0,0 +1,72 @@
+use crate::FlashBlock;
+use reth_chain_state::ExecutedBlock;
+use tracing::{debug, error, info};
+
+/// The `FlashBlockService` maintains an in-memory [`ExecutedBlock`] built out of a sequence of
+/// [`FlashBlock`]s.
+pub trait FlashBlockService {
+    /// An associated error that can occur during execution.
+    type Err;
+
+    /// Adds the `block` into the collection.
+    ///
+    /// Depending on its index and associated block number, it may:
+    /// * Be added to all the flashblocks received prior using this function.
+    /// * Cause a reset of the flashblocks and become the sole member of the collection.
+    /// * Be ignored.
+    fn add_flash_block(&mut self, block: FlashBlock);
+
+    /// Returns the [`ExecutedBlock`] made purely out of [`FlashBlock`]s that were received using
+    /// [`Self::add_flash_block`].
+    fn execute(&mut self) -> Result<ExecutedBlock, Self::Err>;
+}
+
+/// The Reth implementation of [`FlashBlockService`].
+#[derive(Debug)]
+pub struct RethFlashBlockService {
+    current: Option<ExecutedBlock>,
+    blocks: Vec<FlashBlock>,
+}
+
+impl FlashBlockService for RethFlashBlockService {
+    type Err = eyre::Error;
+
+    fn add_flash_block(&mut self, flashblock: FlashBlock) {
```

Comment:
thinking how this should run, I think we'd have

the websocket task, that forwards the messages over a channel to this one, which would be this function effectively, so I think we can maybe replace this either with a `rx: Receiver` or `St: Stream<Item= Flashblock>` generic